### PR TITLE
WIP: Add Bound.ScopeT

### LINKED ISSUE #71: Trouble implementing `Monad` instance when scoping more than one term
Sorry to post this as a GitHub issue, this might not be the best venue for such questions.

I used to have a type like:
```
data Term v
= Constructor (Scope N Term v)
```
which easily satisfied all the good things we could care about (like Monad).

Now, I would like to have a pair of two terms sharing the exact same scope. My initial approach is the following:
```
data Paired f v = Paired (f v, f v)

data Term v
= Constructor (Scope N (Paired Term) v)
```
Unfortunately, the `Paired` type is not a `Monad`, because it makes no sense to substitute variables for pairs of terms. It is, however, an instance of `Bound`, as it supports `Paired f a -> (a -> f c) -> Paired f c`. I cannot figure out how to write the `Monad` instance for `Term` though.

The easy way out is to instead define:
```
data Term v
= Constructor (Scope N f v) (Scope N f v)
```
and manually enforce that the scopes are the same.

But I wonder whether the former approach is doomed, or whether I'm just missing something.

Do all the `Scope` methods that have a requirement `Monad f` really need this requirement, or is it required because it is expected that `f` would be a `Monad`?

### PR DESCRIPTION
**Pull request by phadej**
Fixes #71 

This is still, WIP, but if someone wants to try, be my guest.
- [ ] add more combinators
- [ ] add more instances
- [ ] add Simple variant

### CODE REVIEW

File: src/Bound/ScopeT.hs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    traverse f (ScopeT a) = ScopeT <$> traverse (traverse (traverse f)) a
+
+-- | We cannot write @'Bound' ('ScopeT' n t)@ pre-GHC-8.6 (without an auxiliary type class).
+(>>>>=) :: (Monad f, Functor (t f)) => ScopeT n t f a -> (a -> f b) -> ScopeT n t f b
+ScopeT m >>>>= k = ScopeT $ fmap (fmap (>>= k)) m
+{-# INLINE (>>>>=) #-}
+
+#if __GLASGOW_HASKELL__ >= 805
+-- | @(>>>=) :: ... => 'ScopeT' n t f a -> (a -> f b) -> 'ScopeT' n t f b@
+instance (forall f. Functor f => Functor (t f)) => Bound (ScopeT n t) where
```
Comment (phadej) [inline src/Bound/ScopeT.hs]: I first went with defining

```haskell
class FunctorT t where
    fmapT :: Functor f => (a -> b) -> t f a -> t f b
```

but it was a rabbit hole, `FoldableT`, `TraversableT`, `EqT`, `OrdT`, `ShowT`, `ReadT`...

So I think, better to admit that `QuantifiedConstraints` are good thing, and until GHC-8.6 people can use `>>>>=`.

**Comment by sgraf812** [comment]
Judging from a brief glimpse, this already looks really good to me (if the API is complete wrt. the status quo).

**Comment by ekmett** [comment]
I admit I'm somewhat inclined to split this off as a `bound-extras` or `bound-transformer` style package. It rather drastically increases the level of sophistication required to work with the lib and sort of critically for that judgment doesn't really get to reuse any of the existing functionality of the current `bound` API, so it is only going in here as a matter of stylistic similarity, rather than reuse.

**Comment by phadej** [comment]
@ekmett I can make `bound-transformer` too. I heop adding a mention in the `bound` description / readme (to aid discoverability) is ok?

**Comment by ekmett** [comment]
That'd be more than fine.

**Comment by phadej** [comment]
I made https://github.com/phadej/bound-extras, and will publish it shortly (have to test drive ScopeH a little more).

**Comment by phadej** [comment]
http://hackage.haskell.org/package/bound-extras publshed

**Pull request closed without merging**